pdns/pdns_recursor.cc

```diff
@@ -2786,6 +2786,11 @@ static void daemonize(void)
   }
 }
 
+static void termIntHandler(int)
+{
+  doExitNicely();
```

Comment:
I think it would be better to call `doExit()` instead, I'm worried about race conditions between threads with the use of `exit()`, where one thread tries to access a global structure already destroyed by another, for example.

Reply:
Agree. Calling exit(3) from a signal handler is not safe. 

Even doing stream I/O (which might lead to heap allocations or deallocations) is already dangerous: if the signal is caught while the heap is in an inconsistent state (e.g. inside a new) more heap operations will lead to big trouble.
`doExit()` is already walking a thin rope...

Reply:
Thanks for the comments! I'll do some research and see what I can up with. However, as the main use-case for this fix is Docker, where the container (including all memory/heap allocated to it) will be stopped when the PID 1 (pdns_recursor) stops, this might be not so much of an issue. 

Maybe it would make sense to determine the PID, and only register the handlers if we are PID 1?

Reply:
> Maybe it would make sense to determine the PID, and only register the handlers if we are PID 1?

Please let us not dig us even deeper into a rabbit hole for what is an avoidable Docker bug.

Reply:
oh sigh. Just close it then. Again: it is not a docker bug. It's Linux kernel behaviour: if pid1: you don't get (all) signals. Unless you want to convince Linus to change that behaviour...

Reply:
Ideally we should just set a `sig_atomic_t` flag from the signal handler and call `doExit()` in the main event loop later, but I'm fine with calling `doExit()`. Worst case scenario it causes an abort or a crash, but we were exiting anyway.

Reply:
Agreed. Call doExit() for now. I'll file an issue to do proper shutdown (as rgacogne suggests).

Reply:
@omoerbeek @rgacogne Ok thanks. I'll change the code. Do you prefer I put some "if we are pid 1" logic around the handler registration?